### PR TITLE
Update dataset.register_all to use data_concepts register_all

### PR DESCRIPTION
**Pull request by asantas93**
# Citrine Python PR

## Description 
Improves performance of `dataset.register_all` by using the `register_all` method on the data concepts collection.

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

### Adherence to team decisions
- [ ] I have added tests for 100% coverage
- [ ] I have written Numpy-style docstrings for every method and class.
- [ ] I have communicated the downstream consequences of the PR to others.
- [ ] I have bumped the version in setup.py

### CODE REVIEW

File: src/citrine/resources/dataset.py [hunk 1]
```diff
@@ -252,12 +252,22 @@ def register_all(self, data_concepts_resources: List[ResourceType],
 
         """
         resources = list()
-        for resource in (sorted(data_concepts_resources,
-                                key=lambda resource: writable_sort_order(resource.typ))):
-            registered_resource = self.register(resource, dry_run)
-            if isinstance(registered_resource, BaseEntity):
-                resource.uids = registered_resource.uids
-            resources.append(resource)
+        by_type = {}
```
Comment (maxhutch) [inline src/citrine/resources/dataset.py:255]: You can use a [defaultdict](https://docs.python.org/3.6/library/collections.html#collections.defaultdict) here

File: src/citrine/resources/dataset.py [hunk 2]
```diff
@@ -252,12 +252,22 @@ def register_all(self, data_concepts_resources: List[ResourceType],
...
+        for obj in data_concepts_resources:
+            if obj.typ not in by_type:
+                by_type[obj.typ] = []
+            by_type[obj.typ].append(obj)
+        typ_groups = sorted(list(by_type.values()), key=lambda x: writable_sort_order(x[0]))
+        batch_size = 50
+        for typ_group in typ_groups:
+            num_batches = len(typ_group) // batch_size
+            for batch_num in range(num_batches + 1):
+                batch = typ_group[batch_num: batch_num + batch_size]
```
Comment (maxhutch) [inline src/citrine/resources/dataset.py:265]:
```suggestion
                batch = typ_group[batch_num * batch_size: (batch_num + 1)  * batch_size]
```

Comment (asantas93) [reply src/citrine/resources/dataset.py:265]: oof

**Comment by asantas93** [comment]
> LGTM.
> 
> Do we have e2e test coverage of this method? The fact that the indexing issue still passed tests suggests we're not getting much out of the mocks here, so we should be sure it gets verified.

Maybe one of the table builder tests would fail. I can add a more explicit case for this.

**Pull request merged**